xonsh/prompt/env.py

```diff
@@ -1,39 +1,55 @@
 """Prompt formatter for virtualenv and others"""
 
-import os
+import re
+from pathlib import Path
 
 from xonsh.built_ins import XSH
 
 
-def find_env_name():
-    """Finds the current environment name from $VIRTUAL_ENV or
-    $CONDA_DEFAULT_ENV if that is set.
+def env_name():
+    """Build env_name based on different sources. Respect order of precedence.
+
+    Name from VIRTUAL_ENV_PROMPT will be used as-is.
+    Names from other sources are surrounded with ``{env_prefix}`` and
+    ``{env_postfix}`` fields.
     """
-    env_path = XSH.env.get("VIRTUAL_ENV", "")
-    if env_path:
-        env_name = os.path.basename(env_path)
-    else:
-        env_name = XSH.env.get("CONDA_DEFAULT_ENV", "")
-    return env_name
+    if XSH.env.get("VIRTUAL_ENV_DISABLE_PROMPT"):
+        return
+    virtual_env_prompt = XSH.env.get("VIRTUAL_ENV_PROMPT")
+    if virtual_env_prompt:
+        return virtual_env_prompt
+    virtual_env = XSH.env.get("VIRTUAL_ENV")
+    if virtual_env:
+        venv_path = Path(virtual_env)
+        pyvenv_cfg_prompt = prompt_from_pyvenv_cfg(venv_path)
+        if pyvenv_cfg_prompt:
```

Comment:
we can use caching here as well. with path as the key using lru_cache with size as 3-5

Reply:
See my comment above. This code path wouldn't even be hit anymore then.

Reply:
ah ... I think this is what you meant, I said the same thing above :D - will have a go at this now. Thanks!